diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 300e1f7..11cc02e 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -72,7 +72,7 @@ set_target_properties(negf PROPERTIES Fortran_MODULE_DIRECTORY ${BUILD_MOD_DIR})
 
 if(WITH_MPI)
   find_package(MPI REQUIRED)
-  target_link_libraries(negf PUBLIC MpiFx::MpiFx)
+  target_link_libraries(negf PUBLIC MpiFx::MpiFx MPI::MPI_Fortran)
 endif()
 target_link_libraries(negf PRIVATE ${BLAS_LIBRARIES} ${LAPACK_LIBRARIES})
 
diff --git a/src/mpi_globals.F90 b/src/mpi_globals.F90
index 8009cd7..de140f3 100644
--- a/src/mpi_globals.F90
+++ b/src/mpi_globals.F90
@@ -24,7 +24,8 @@ module mpi_globals
 
 #:if defined("MPI")
 
-  use libmpifx_module
+  use mpi, only : mpi_cart_create, mpi_cart_sub
+  use libmpifx_module, only : mpifx_comm
 
 #:endif
 
@@ -61,14 +62,18 @@ subroutine negf_cart_init(inComm, nk, cartComm, energyComm, kComm)
 
       type(mpifx_comm) :: cartComm
       integer :: outComm
-      integer :: ndims = 2
+      integer :: ndims
       integer :: dims(2)
-      integer :: period(2) = 0
-      integer :: remain_dims(2)
+      logical :: periods(2)
+      logical :: remain_dims(2)
       integer :: nE
-      integer :: reorder = 1
+      logical :: reorder
       integer :: mpierr
 
+      ndims = 2
+      periods(:) = .false.
+      reorder = .true.
+
       if (mod(inComm%size,nk) /=0 ) then
         stop "Error in cart_init: cannot build a 2D cartesian grid with incompatible sizes"
       end if
@@ -78,11 +83,11 @@ subroutine negf_cart_init(inComm, nk, cartComm, energyComm, kComm)
       call MPI_CART_CREATE(inComm%id, ndims, dims, periods, reorder, outComm, mpierr)
       call cartComm%init(outComm, mpierr)
 
-      remain_dims = (/0, 1/)
+      remain_dims(:) = [.false., .true.]
       call MPI_CART_SUB(cartComm%id, remain_dims, outComm, mpierr)
       call energyComm%init(outComm, mpierr)
 
-      remain_dims = (/1, 0/)
+      remain_dims(:) = [.true., .false.]
       call MPI_CART_SUB(cartComm%id, remain_dims, outComm, mpierr)
       call kComm%init(outComm, mpierr)
 

